### PR TITLE
Add unit testing

### LINKED ISSUE #47: Add unit testing
Experience with #46 has demonstrated that confirming consistent behavior on Linux, Mac, and Windows is tricky even aside from compiler issues. (even if something does compile for all platforms) What appears to work on one platform may not in work on another. This has only further reinforced a need for unit testing run through the CI on all three systems for the dataloader. This way even if something compiles on all major platforms, a single developer can also confirm that it will perform exactly as expected on all major platforms as well, instead of just presuming that it works because it compiled, having to rely on multiple developers running multiple platforms, or running things in a VM.

### PR DESCRIPTION
**Pull request by Spartan322**
Fixes #47 

Make github action tests run explicit
Fix dropping annotation list for Errors
Fix potential empty get_errors crashes
Fix incorrect csv error behavior
Add use_sep for `LineObject` and `std::vector<LineObject>`
Remove constexpr of load_from_buffer and load_from_string for parsers

Add https://github.com/snitch-org/snitch/commit/d6632123cc8d13bdbc5cd60fd6741b9e0f635e82
Make versioned submodules ignore dirty
Add tests/bin/* to gitignore

### CODE REVIEW

File: src/openvic-dataloader/csv/Parser.cpp [hunk 1]
```diff
@@ -166,7 +176,7 @@ bool Parser::parse_csv(bool handle_strings) {
 	}();
 	_has_error = _parse_handler->parse_state().logger().errored();
 	_has_warning = _parse_handler->parse_state().logger().warned();
-	if (!errors->empty()) {
+	if (errors->empty()) {
```
Comment (wvpm) [inline src/openvic-dataloader/csv/Parser.cpp:179]: This change doesn't make sense to me.
If errors are empty, _has_error should be false.
So the original was correct.

File: tests/src/Error.cpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+		CHECK(prim_annotation->message() == "primary annotation"sv);
+	}
+
+	auto* sec_annotation = errors.create<SecondaryAnnotation>("secondary annotation");
+	CHECK_IF(sec_annotation) {
+		CHECK(sec_annotation->kind() == ErrorKind::SecondaryAnnotation);
+		CHECK(sec_annotation->message() == "secondary annotation"sv);
+	}
+
+	if (!prim_annotation || !sec_annotation) {
```
Comment (wvpm) [inline tests/src/Error.cpp:148]: Could you explain this part?
return statements in tests are highly unusual.

Comment (Spartan322) [reply tests/src/Error.cpp]: Eh I guess its not a big deal, in this specific case it'll repeatedly complain about the annotations being null without the return.

File: tests/src/csv/LineObject.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+#include <openvic-dataloader/csv/LineObject.hpp>
+
+#include <range/v3/range/primitives.hpp>
+#include <snitch/snitch.hpp>
+
+using namespace ovdl;
+using namespace csv;
+using namespace std::string_view_literals;
+
+TEST_CASE("LineObject", "[line-object]") {
```
Comment (wvpm) [inline tests/src/csv/LineObject.cpp:13]: Nice tests. Once I adjusted to the format, I quickly understood what each test does.

Comment (Spartan322) [reply tests/src/csv/LineObject.cpp:13]: Yeah snitch is a modern C++20 testing suite based on some of the interface to Catch2.

**Review by wvpm — APPROVED** [review]

**Pull request merged**